storm-core/src/jvm/backtype/storm/blobstore/NimbusBlobStore.java

```diff
@@ -0,0 +1,419 @@
+      this.eof = (listBlobs.get_keys_size() == 0);
+    }
+    
+    private boolean isCacheEmpty() {
+      return listBlobs.get_keys_size() <= offset;
+    }
+
+    private void readMore() {
+      if (!eof) {
+        try {
+          offset = 0;
+          synchronized(client) {
+            listBlobs = client.getClient().listBlobs(listBlobs.get_session());
+          }
+          if (listBlobs.get_keys_size() == 0) {
+            eof = true;
+          }
+        } catch (TException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    
+    @Override
+    public synchronized boolean hasNext() {
+      if (isCacheEmpty()) {
+        readMore();
+      }
+      return !eof;
+    }
+
+    @Override
+    public synchronized String next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      String ret = listBlobs.get_keys().get(offset);
+      offset++;
+      return ret;
+    }
+    
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException("Delete Not Supported");
+    }
+  }
+
+  public class NimbusDownloadInputStream extends InputStreamWithMeta {
+    private BeginDownloadResult beginBlobDownload;
+    private byte[] buffer = null;
+    private int offset = 0;
+    private int end = 0;
+    private boolean eof = false;
+
+    public NimbusDownloadInputStream(BeginDownloadResult beginBlobDownload) {
+      this.beginBlobDownload = beginBlobDownload;
+    }
+
+    @Override
+    public long getVersion() throws IOException {
+      return beginBlobDownload.get_version();
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+      if (isEmpty()) {
+        readMore();
+        if (eof) {
+          return -1;
+        }
+      }
+      int length = Math.min(1, available());
+      if (length == 0) {
+        return -1;
+      }
+      int ret = buffer[offset];
+      offset += length;
+      return ret;
+    }
+    
+    @Override 
+    public synchronized int read(byte[] b, int off, int len) throws IOException {
+      if (isEmpty()) {
+        readMore();
+        if (eof) {
+          return -1;
+        }
+      }
+      int length = Math.min(len, available());
+      System.arraycopy(buffer, offset, b, off, length);
+      offset += length;
+      return length;
+    }
+    
+    private boolean isEmpty() {
+      return buffer == null || offset >= end;
+    }
+    
+    private void readMore() {
+      if (!eof) {
+        try {
+          ByteBuffer buff;
+          synchronized(client) {
+            buff = client.getClient().downloadBlobChunk(beginBlobDownload.get_session());
+          }
+          buffer = buff.array();
+          offset = buff.arrayOffset() + buff.position();
+          int length = buff.remaining();
+          end = offset + length;
+          if (length == 0) {
+            eof = true;
+          }
+        } catch (TException e) {
+          throw new RuntimeException(e);
```

Comment:
Is it better to just throw TException, instead of RuntimeException? Let caller to decide what the further action for TExcetpion, since the latter one will possibly cause the thread of caller crash. e.g.  for supervisor, we have a retry mechanism in Utils/downloadResourcesAsSupervisor. If RuntimeException, the retry might not work. Same issue for other operations in this file.